src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -581,6 +628,130 @@ fn max_slice_length<'p, 'a: 'p, 'tcx: 'a, I>(
+                    I128  => offset_sign_for_ty!(i128, u128),
+                }
+            }
+            ty::TyUint(_) | ty::TyChar => {
+                (lo, hi)
+            }
+            _ => bug!("`Interval` should only contain integer types")
+        }
+    }
+
+    fn into_inner(self) -> (u128, u128) {
+        (self.lo, self.hi)
+    }
+}
+
+/// Given a pattern in a `match` and a collection of ranges corresponding to the
+/// domain of values of a type (say, an integer), return a new collection of ranges
+/// corresponding to those ranges minus the ranges covered by the pattern.
+fn ranges_subtract_pattern<'a, 'tcx>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
+                                     pat_ctor: &Constructor<'tcx>,
+                                     ranges: Vec<Constructor<'tcx>>)
+                                     -> Vec<Constructor<'tcx>> {
+    if let Some(pat_interval) = Interval::from_ctor(pat_ctor) {
+        let mut remaining_ranges = vec![];
+        let mut ranges: Vec<_> = ranges.into_iter().filter_map(|r| {
+            Interval::from_ctor(&r).map(|i| i.into_inner())
+        }).collect();
+        for (subrange_lo, subrange_hi) in ranges {
+            if pat_interval.lo > subrange_hi || pat_interval.hi < subrange_lo {
+                // The pattern doesn't intersect with the subrange at all,
+                // so the subrange remains untouched.
+                remaining_ranges.push((subrange_lo, subrange_hi));
+            } else if pat_interval.lo <= subrange_lo && pat_interval.hi >= subrange_hi {
+                // The pattern entirely covers the subrange of values,
+                // so we no longer have to consider this subrange_
+            } else if pat_interval.lo <= subrange_lo {
+                // The pattern intersects a lower section of the subrange,
+                // so only the upper section will remain.
+                remaining_ranges.push((pat_interval.hi + 1, subrange_hi));
+            } else if pat_interval.hi >= subrange_hi {
+                // The pattern intersects an upper section of the subrange,
+                // so only the lower section will remain.
+                remaining_ranges.push((subrange_lo, pat_interval.lo - 1));
+            } else {
+                // The pattern intersects the middle of the subrange,
+                // so we create two ranges either side of the intersection.)
+                remaining_ranges.push((subrange_lo, pat_interval.lo));
+                remaining_ranges.push((pat_interval.hi, subrange_hi));
+            }
+        }
+        // Convert the remaining ranges from pairs to inclusive `ConstantRange`s.
+        let ty = pat_interval.ty;
+        remaining_ranges.into_iter().map(|(lo, hi)| {
+            let (lo, hi) = Interval::offset_sign(ty, (lo, hi), false);
+            ConstantRange(ty::Const::from_bits(cx.tcx, lo, ty),
+                          ty::Const::from_bits(cx.tcx, hi, ty),
```

Comment:
Should `ConstantRange` use `ty::Const`? Seems expensive, and useless unless the `ty::Const` is `Bits` (cc @oli-obk).

Reply:
We should definitely be using `ConstValue` instead of `ty::Const`. I put it on my TODO list

Reply:
`ConstantRange` can only hold integers, so `ConstValue` shouldn't be needed, right?

Reply:
Probably. My TODO already says "ConstValue or smaller" ;). That should not block this PR though, since it's a preexisting issue.